include/freerdp/codec/h264.h

```diff
@@ -96,9 +97,25 @@ extern "C"
 	 * @param pDstSize A pointer for the destination buffer size in bytes
 	 * @return \b >= 0 for success, \b <0 for an error
 	 * @since version 3.6.0
+	 * @deprecated use h264_compress_ex instead
 	 */
 	FREERDP_API INT32 h264_compress(H264_CONTEXT* h264, BYTE** ppDstData, UINT32* pDstSize);
 
+	/**
+	 * @brief Compress directly supplied image data to H264 stream
+	 *
+	 * @param h264 The H264 context to use for compression
+	 * @param pSrcData A pointer to pointers to input buffers (in YUV420P or NV12 format)
+	 * @param pSrcStride A pointer to array of line byte lengths in the input buffers
+	 * @param ppDstData A pointer that will hold the allocated result buffer
+	 * @param pDstSize A pointer for the destination buffer size in bytes
+	 * @return \b >= 0 for success, \b <0 for an error
+	 * @since version
+	 */
+	FREERDP_API INT32 h264_compress_ex(H264_CONTEXT* h264, const BYTE** WINPR_RESTRICT pSrcData,
```

Comment:
so, like mentioned before:
I´d suggest moving format conversion from source format to required internal FFMPEG format from channel into this function.
this way we do not need to link the `rdpecam` channel to `ffmpeg` nor `swscale` and can freely implement another backend if we find something better.

only downside is that you need to duplicate the `CAM_MEDIA_FORMAT` enum here to allow these as input formats.

NOTE: would also move the `MJPEG` encoding to such a function

Reply:
@akallabeth : I liked your suggestion to add source raw format type to `h264_compress_ex`, but I'm not sure converting from MJPEG to raw format should also move inside `h264.c`, because it is a separate step not directly related  to h264 encoding. Also I'm not sure `CAM_MEDIA_FORMAT` is the right type to pass this new parameter. I would probably use `enum AVPixelFormat` for that, which is part of `swscale` library and has many more types (including types produced by ffmpeg MJPEG decoder).
Anyway, let me take some time to think about it more and try to come up with something.

Reply:
@oleg0421 `AVPixelFormat` is not suitable as that is defined in an optional dependent library.

Reply:
@akallabeth: supporting runtime fallback to software encoder (in case hardware encoder initialization fails) requires  changes of multiple layers, including `h264_ffmpeg.c` and `h264_openh264.c` subsystems, common `h264.c` layer and camera channel's `encoding.c` layer. Unfortunately, I don't currently have enough time to work on that. The `CHANNEL_RDPECAM_CLIENT` is not enabled by default anyway, users need to compile the code to enable it. They might as well try `WITH_VAAPI_H264_ENCODING` option and if it doesn't work on their system, disable it.

Reply:
does not necessarily require changing all the backends, but as we´re chaning the required `API` to support the channel we should address this well enough so that we don´t need to rework this in a few months.

the point being, if you have the `API` you can add the implementation for others later on.

Reply:
@oleg0421 just to be clear: I just want the `API` function `h264_compress_ex` to cover all these possibilities as changing the implementation is something we can easily do, but changing function signatures is not.

Reply:
@akallabeth : I understand your point and have an idea how to support runtime fallback to software encoding:

1. Move `avcodec_find_encoder_by_name("h264_vaapi")` and `av_hwdevice_ctx_create()` to `libavcodec_init()`. If they fail, switch to using software encoder.
2. In other places, rely on `if (sys->hwctx)` check to support both encoders in runtime.
3. Add "get only" option `H264_CONTEXT_OPTION_ENCODER_INPUT_FORMAT` to `H264_CONTEXT_OPTION` enum with these values:
```
typedef enum
{
	H264_ENCODER_INPUT_FORMAT_YUV420P = 0,
	H264_ENCODER_INPUT_FORMAT_NV12,
} H264_ENCODER_INPUT_FORMAT;
```

4. Before using `h264_compress_ex()` API with proposed signature, call:
```
h264_context_get_option(stream->h264, H264_CONTEXT_OPTION_ENCODER_INPUT_FORMAT);
```
which will return the right format, depending on which encoder was initialized, and supply `pSrcData` in right format for `h264_compress_ex()`.

This way, openh264 library will also be supported, because it will always return YUV420P.

Reply:
@oleg0421 nice idea with the fallback.
as for formats, I´d take the opposite approach, make the internals opaque for the caller and just tell which format the input uses.
This way you can switch internal formats as much as you like if the backend has better support for a different one without the need to modify the caller code.

Reply:
> @akallabeth : I understand your point and have an idea how to support runtime fallback to software encoding:
> 
>     1. Move `avcodec_find_encoder_by_name("h264_vaapi")` and `av_hwdevice_ctx_create()` to `libavcodec_init()`. If they fail, switch to using software encoder.
> 
>     2. In other places, rely on `if (sys->hwctx)` check to support both encoders in runtime.
> 

That was exactly what I was meaning with fallback (BTW I already have implemented that in old private projects, it works great)

Reply:
> [note] the whole `sws_scale` stuff you do you can just ditch and the current (implicit) input expectation is `H264_ENCODER_INPUT_FORMAT_YUV420P` - which works with openh264 as well. for `ffmpeg` you can just do format conversion internally without any modifications to the `S_H264_CONTEXT_SUBSYSTEM` struct

We can only move `sws_scale()` to `h264.c`, but not ditch it, in case we add format type parameter to `h264_compress_ex()`. Some layer still needs to do the conversion from YUYV422 (which cameras typically output) or another camera format to YUV420P or NV12.

We cannot put `sws_scale()` logic inside `h264_ffmpeg.c`, because `h264_openh264.c` needs same logic to convert from camera output  format to YUV420P.

The h264 encoders typically expect YUV420P input for software and NV12 for hardware encoding. It's not typically h264 encoder library responsibility to handle variety of input formats. Also sometimes video processing does up- or down-scaling before encoding, and `sws_scale()` efficiently does it together with format conversion, which is the reason to keep it at current place (`encoding.c`).

Reply:
@oleg0421 please forget about openh264.
we can easily move it inside and have a `ffmpeg` implementation in the function as first step.
what you are talking about is not required for this pr and out of scope, but no reason not to clean that up.

the drawbacks of having it outside is that you need to not only  link the libraries, you need to know which backend is compiled in, ....

for scaling we have the `freerdp_image_scale` function and optimized functions are something out of scope here as well (and they again must not expose implementation detail to be switchable)

Reply:
@akallabeth : what about MJPEG decoding part? Also having it  inside `h264_ffmpeg.c` makes little sense. You are thinking more in terms of ffmpeg library dependencies, but we also need to keep modularity in mind. `h264.c` is the unified interface to different h264 encoder/decoder libraries, not a container for everything ffmpeg related.
If you put everything in `h264_ffmpeg.c` now, then we'll have to duplicate the logic later for other h264 encoders.

`freerdp_image_scale` doesn't support any of those YUV formats, only RGB-based, so we need `sw_scale`.

Reply:
well, you have only two paths:
1. pass  through directly
2. convert to h264

since that is just another input format it can be converted internally

> `freerdp_image_scale` doesn't support any of those YUV formats, only RGB-based, so we need sw_scale.

no, this just illustrates that we have functions abstracting the required functions and (just as your example) the things are currently not implemented due to lack of demand.
not an argument to reinvent the wheel.

Reply:
@akallabeth : MJPEG is not a raw video format and can't be represented in terms of `BYTE* srcSlice[4]` and `srcLineSizes[4]`, so if you want to support it by `h264_compress_ex`, we need to change parameters to point to whole buffer and provide whole image width / height. Only this way we can get rid of channel's `swscale` and `ffmpeg` libraries dependency, but at the cost of polluting h264 encoder/decoder library and transferring the above dependencies there.

Reply:
@oleg0421 just like `avc444_compress` then?
that already does image from `RGB` to 2x YUV420 conversion.

Reply:
@oleg0421 note: encoding/decoding is split:
* `libfreerdp/codec/h264.c` for generic code
* `libfreerdp/codec/h264_ffmpeg.c` for the ffmpeg implementation details 
* ...

Reply:
> @oleg0421 just like `avc444_compress` then? that already does image from `RGB` to 2x YUV420 conversion.

@akallabeth these functions use primitives such as `ssse3_RGBToYUV420_BGRX_UV`, which use manually written SSE3 optimized instructions. This is beyond my expertise and besides `sws_scale()` does these optimized conversions very efficiently.